### ISSUE TITLE
Generic inference from LHS of assignment when destructuring can lead to unexpected `any`s 

### ISSUE THREAD

**Issue by jkillian** [microsoft/TypeScript#45074]
# Bug Report

### 🔎 Search Terms

Generic inference, destructuring, binding patterns

### 🕗 Version & Regression Information

Exists in 4.3.5

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about generics

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.3.5&ssl=3&ssc=24&pln=1&pc=1&strict=true#code/CYUwxgNghgTiAEAzArgOzAFwJYHtXwHMRUQYswAxNMAHgBUA+ACgEoAueOgbgCgw8AzhngBveACNY8AL7wAvIWKlyVdK14B6DfAAGkmDvhYBuqKgCehgBakQAOh5A)

### 💻 Code

```ts
declare function genericFunc<T>(): T;
const { bar } = genericFunc();
// `bar` is `any` here.
```

### 🙁 Actual behavior

`bar` is `any` and `T` is inferred to be `{ bar: any }`.

### 🙂 Expected behavior

The inferred type of `T` should be `unknown` and there should be an error from trying to destructure it.
(Or, `T` could still be inferred as `{ bar: any }` but then `noImplicitAny` should trigger an error. This seems at a glance less desirable to me than the former, but I don't know enough how generic inference works to make a more informed recommendation. Anyways, some sort of error no matter how it happens is what would be ideal!)

Notes: see https://github.com/microsoft/TypeScript/issues/43371#issuecomment-881641347 for the original context.

**Comment by andrewbranch** [comment]
The inference of `{ bar: any }` may or may not be avoidable without breaking a lot of other inferences, but it should at least error under `noImplicitAny`.

**Comment by andrewbranch** [comment]
Related: https://github.com/microsoft/TypeScript/issues/43605

**Comment by jkillian** [comment]
Related to the discussion in https://github.com/microsoft/TypeScript/issues/46012 I believe

**Comment by andrewbranch** [comment]
Yeah. Probably not going to fix the `noImplicitAny` error for 4.5 because I want to investigate making this whole pattern an error in the 4.6 timeframe, as discussed in that design meeting.

**Comment by jkillian** [comment]
Love that this is fixed, thanks @andrewbranch!
